File: packages/protocol/src/stream.ts
/**
 * The dictation WebSocket protocol.
 *
 * One socket carries both control and audio: JSON text frames for control,
 * binary frames for raw `linear16` PCM (see ./audio.ts). Binary frames are
 * only meaningful between a `start` and a `stop`/`cancel`.
 *
 * A session is one utterance. The client opens the socket on hotkey-down and
 * closes it after the `result` arrives, so TLS setup and Durable Object
 * wake-up overlap with the user's first words instead of adding to latency.
 */

/** Sent by the client to open a dictation session. Must precede any audio. */
export interface StartFrame {
  type: "start";
  /** Must equal SAMPLE_RATE; sent explicitly so the server can reject a mismatch. */
  sampleRate: number;
  encoding: "linear16";
  /** BCP-47-ish language hint. Omit to let the model detect. */
  locale?: string;
  /**
   * Active organization. The server independently verifies the caller is a
   * member of this org; a client-supplied value is never trusted on its own.
   */
  orgId?: string | null;
  /**
   * Vocabulary boosts — the user's personal terms merged with their org's
   * shared glossary. Passed to the recognizer and reused as spelling context
   * for the cleanup pass.
   */
  keyterms?: string[];
  /**
   * Focused application, for history and as a cleanup style hint (code
   * editor, email, chat). Never used for routing or authorization.
   */
  appName?: string | null;
  /** Run the cleanup pass. When false the raw transcript is returned as-is. */
  format?: boolean;
  /**
   * Store this transcript in the user's history. The person can only opt
   * out: an org that turned retention off still wins when this is true.
   */
  retain?: boolean;
}

/**
 * What surrounds the cursor in the field being dictated into, read when the
 * hotkey went down. Wispr Flow–style context: cleanup uses it to continue a
 * sentence, spell names already on screen, and tell a Gmail tab from a
 * Slack one. Used for this dictation's cleanup only; never stored.
 */
export interface FieldContext {
  /** Text just before the cursor, at most MAX_CONTEXT_BEFORE characters. */
  before?: string;
  /** Text just after the cursor, at most MAX_CONTEXT_AFTER characters. */
  after?: string;
  /** Focused window title — in a browser it names the site ("Inbox - Gmail"). */
  windowTitle?: string;
}

export const MAX_CONTEXT_BEFORE = 1_500;
export const MAX_CONTEXT_AFTER = 500;
export const MAX_WINDOW_TITLE = 200;

/** Sent on hotkey release. The server finalizes, cleans up, and replies `result`. */
export interface StopFrame {
  type: "stop";
  /**
   * Cursor context, captured at hotkey-down and sent here because it is only
   * needed for cleanup. Servers that predate it ignore the field.
   */
  context?: FieldContext;
}

/** Sent on Escape. The server discards the utterance and emits no `result`. */
export interface CancelFrame {
  type: "cancel";
}

/** Keeps intermediaries from idling the socket out during a long pause. */
export interface PingFrame {
  type: "ping";
}

export type ClientFrame = StartFrame | StopFrame | CancelFrame | PingFrame;

/** Server accepted the `start` and is ready for audio. */
export interface ReadyEvent {
  type: "ready";
  sessionId: string;
}

/**
 * Interim recognition result. Overwrites the previous partial rather than
 * appending. Display-only — never inject a partial, it will be revised.
 */
export interface PartialEvent {
  type: "partial";
  text: string;
}

/** The recognizer's final, before cleanup. Emitted even when formatting runs. */
export interface TranscriptEvent {
  type: "transcript";
  text: string;
}

/** Terminal success. This is the text to inject. */
export interface ResultEvent {
  type: "result";
  /** Text to inject: cleaned when `formatted`, otherwise identical to `raw`. */
  text: string;
  /** The recognizer's output before cleanup, retained for history and debugging. */
  raw: string;
  /**
   * Whether cleanup actually ran. False when disabled, when it failed, or when
   * it exceeded its deadline and the raw transcript was shipped instead.
   */
  formatted: boolean;
  /** Audio duration, used for metering. */
  durationMs: number;
}

export type ErrorCode =
  | "unauthorized"
  | "org_forbidden"
  | "quota_exceeded"
  | "bad_request"
  | "upstream_failed"
  | "internal";

/** Terminal failure. `retryable` distinguishes a blip from a wasted retry. */
export interface ErrorEvent {
  type: "error";
  code: ErrorCode;
  message: string;
  retryable: boolean;
}

export interface PongEvent {
  type: "pong";
}

export type ServerEvent =
  | ReadyEvent
  | PartialEvent
  | TranscriptEvent
  | ResultEvent
  | ErrorEvent
  | PongEvent;

/**
 * The JWT is passed as a WebSocket subprotocol rather than a header.
 *
 * Browsers cannot set headers on a WebSocket handshake. A Rust client can, but
 * using the subprotocol keeps a future browser client possible without a
 * second auth path on the server.
 */
export const WS_SUBPROTOCOL_PREFIX = "weldspeak.token.";

/** Build the subprotocol value carrying `token`. */
export function encodeTokenSubprotocol(token: string): string {
  return `${WS_SUBPROTOCOL_PREFIX}${token}`;
}

/** Recover a token from a `Sec-WebSocket-Protocol` list, or null if absent. */
export function decodeTokenSubprotocol(header: string | null): string | null {
  if (!header) return null;
  for (const raw of header.split(",")) {
    const value = raw.trim();
    if (value.startsWith(WS_SUBPROTOCOL_PREFIX)) {
      const token = value.slice(WS_SUBPROTOCOL_PREFIX.length);
      return token.length > 0 ? token : null;
    }
  }
  return null;
}

/**
 * Narrow an untrusted parsed JSON value to a ClientFrame.
 *
 * Frames arrive from the network, so this validates shape rather than
 * asserting it. Returns null on anything unrecognized; callers reply with a
 * `bad_request` error rather than throwing.
 */
/**
 * Validate cursor context from the client.
 *
 * Clamped rather than rejected: context is a hint, and an oversized field
 * from a large document should cost a trim, not the dictation. The text
 * nearest the cursor is what matters, so `before` keeps its end and `after`
 * its start.
 */
export function parseFieldContext(value: unknown): FieldContext | undefined {
  if (typeof value !== "object" || value === null) return undefined;
  const raw = value as Record<string, unknown>;
  const text = (field: unknown) => (typeof field === "string" && field.trim() ? field : undefined);

  const before = text(raw.before)?.slice(-MAX_CONTEXT_BEFORE);
  const after = text(raw.after)?.slice(0, MAX_CONTEXT_AFTER);
  const windowTitle = text(raw.windowTitle)?.trim().slice(0, MAX_WINDOW_TITLE);

  if (!before && !after && !windowTitle) return undefined;
  return {
    ...(before ? { before } : {}),
    ...(after ? { after } : {}),
    ...(windowTitle ? { windowTitle } : {}),
  };
}

export function parseClientFrame(value: unknown): ClientFrame | null {
  if (typeof value !== "object" || value === null) return null;
  const frame = value as Record<string, unknown>;

  switch (frame.type) {
    case "stop": {
      const context = parseFieldContext(frame.context);
      return context ? { type: "stop", context } : { type: "stop" };
    }
    case "cancel":
      return { type: "cancel" };
    case "ping":
      return { type: "ping" };
    case "start": {
      if (typeof frame.sampleRate !== "number") return null;
      if (frame.encoding !== "linear16") return null;

      const keyterms = Array.isArray(frame.keyterms)
        ? frame.keyterms.filter((t): t is string => typeof t === "string")
        : undefined;

      return {
        type: "start",
        sampleRate: frame.sampleRate,
        encoding: "linear16",
        locale: typeof frame.locale === "string" ? frame.locale : undefined,
        orgId: typeof frame.orgId === "string" ? frame.orgId : null,
        keyterms,
        appName: typeof frame.appName === "string" ? frame.appName : null,
        format: typeof frame.format === "boolean" ? frame.format : true,
        retain: typeof frame.retain === "boolean" ? frame.retain : true,
      };
    }
    default:
      return null;
  }
}

File: packages/protocol-rs/src/stream.rs
//! The dictation WebSocket protocol.
//!
//! Mirrors `packages/protocol/src/stream.ts`. One socket carries JSON text
//! frames for control and binary frames for raw `linear16` PCM. Binary frames
//! are only meaningful between a `start` and a `stop`/`cancel`.

use serde::{Deserialize, Serialize};

/// Frames sent by the desktop client.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ClientFrame {
    /// Opens a dictation session. Must precede any audio.
    #[serde(rename = "start", rename_all = "camelCase")]
    Start {
        sample_rate: u32,
        encoding: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        locale: Option<String>,
        /// Active organization. The server re-verifies membership; a
        /// client-supplied value is never trusted on its own.
        #[serde(skip_serializing_if = "Option::is_none")]
        org_id: Option<String>,
        /// Vocabulary boosts: personal terms merged with the org glossary.
        #[serde(skip_serializing_if = "Option::is_none")]
        keyterms: Option<Vec<String>>,
        /// Focused application, for history and diagnostics only.
        #[serde(skip_serializing_if = "Option::is_none")]
        app_name: Option<String>,
        /// Run the cleanup pass. When false the raw transcript is returned.
        #[serde(skip_serializing_if = "Option::is_none")]
        format: Option<bool>,
        /// Store the transcript in history. Only `false` changes anything:
        /// the server already defaults to keeping it unless the org says not.
        #[serde(skip_serializing_if = "Option::is_none")]
        retain: Option<bool>,
    },
    /// Sent on hotkey release. The server finalizes and replies `result`.
    #[serde(rename = "stop")]
    Stop,
    /// Sent on Escape. The server discards the utterance; no `result` follows.
    #[serde(rename = "cancel")]
    Cancel,
    #[serde(rename = "ping")]
    Ping,
}

/// Why a session failed. `retryable` on [`ServerEvent::Error`] says whether a
/// retry is worth attempting.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ErrorCode {
    Unauthorized,
    OrgForbidden,
    QuotaExceeded,
    BadRequest,
    UpstreamFailed,
    Internal,
}

/// Events sent by the Worker.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ServerEvent {
    /// The `start` was accepted; the server is ready for audio.
    #[serde(rename = "ready", rename_all = "camelCase")]
    Ready { session_id: String },

    /// Interim result. Replaces the previous partial rather than appending.
    /// Display-only — never inject a partial, it will be revised.
    #[serde(rename = "partial")]
    Partial { text: String },

    /// The recognizer's final, before cleanup.
    #[serde(rename = "transcript")]
    Transcript { text: String },

    /// Terminal success. `text` is what gets injected.
    #[serde(rename = "result", rename_all = "camelCase")]
    Result {
        text: String,
        raw: String,
        /// False when cleanup was disabled, failed, or missed its deadline.
        formatted: bool,
        duration_ms: u64,
    },

    /// Terminal failure.
    #[serde(rename = "error")]
    Error {
        code: ErrorCode,
        message: String,
        retryable: bool,
    },

    #[serde(rename = "pong")]
    Pong,
}

/// The JWT travels as a WebSocket subprotocol rather than a header.
///
/// Browsers cannot set headers on a WebSocket handshake. This client could,
/// but using the subprotocol keeps a future browser client possible without a
/// second auth path on the server.
pub const WS_SUBPROTOCOL_PREFIX: &str = "weldspeak.token.";

/// Build the subprotocol value carrying `token`.
pub fn encode_token_subprotocol(token: &str) -> String {
    format!("{WS_SUBPROTOCOL_PREFIX}{token}")
}

/// Recover a token from a `Sec-WebSocket-Protocol` list, if present.
pub fn decode_token_subprotocol(header: &str) -> Option<&str> {
    header.split(',').find_map(|raw| {
        raw.trim()
            .strip_prefix(WS_SUBPROTOCOL_PREFIX)
            .filter(|token| !token.is_empty())
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn start_frame_serializes_to_the_typescript_shape() {
        let frame = ClientFrame::Start {
            sample_rate: 16_000,
            encoding: "linear16".into(),
            locale: Some("en".into()),
            org_id: Some("org_123".into()),
            keyterms: Some(vec!["Inconel 625".into()]),
            app_name: None,
            format: Some(true),
            retain: Some(false),
        };

        let json: serde_json::Value = serde_json::to_value(&frame).unwrap();
        assert_eq!(json["type"], "start");
        assert_eq!(json["sampleRate"], 16_000);
        assert_eq!(json["orgId"], "org_123");
        assert_eq!(json["keyterms"][0], "Inconel 625");
        assert_eq!(json["retain"], false);
        // Absent optionals are omitted rather than sent as null.
        assert!(json.get("appName").is_none());
    }

    #[test]
    fn unit_frames_carry_only_a_type() {
        assert_eq!(
            serde_json::to_string(&ClientFrame::Stop).unwrap(),
            r#"{"type":"stop"}"#
        );
    }

    #[test]
    fn server_events_deserialize_from_the_typescript_shape() {
        let event: ServerEvent = serde_json::from_str(
            r#"{"type":"result","text":"Hello.","raw":"hello","formatted":true,"durationMs":1200}"#,
        )
        .unwrap();

        match event {
            ServerEvent::Result {
                text,
                formatted,
                duration_ms,
                ..
            } => {
                assert_eq!(text, "Hello.");
                assert!(formatted);
                assert_eq!(duration_ms, 1200);
            }
            other => panic!("expected Result, got {other:?}"),
        }
    }

    #[test]
    fn error_codes_use_snake_case_on_the_wire() {
        let event: ServerEvent = serde_json::from_str(
            r#"{"type":"error","code":"quota_exceeded","message":"cap reached","retryable":false}"#,
        )
        .unwrap();

        match event {
            ServerEvent::Error { code, .. } => assert_eq!(code, ErrorCode::QuotaExceeded),
            other => panic!("expected Error, got {other:?}"),
        }
    }

    #[test]
    fn token_survives_a_subprotocol_round_trip() {
        let encoded = encode_token_subprotocol("abc.def.ghi");
        assert_eq!(decode_token_subprotocol(&encoded), Some("abc.def.ghi"));
    }

    #[test]
    fn token_is_found_among_other_offered_subprotocols() {
        let header = format!("chat, {}, superchat", encode_token_subprotocol("tok"));
        assert_eq!(decode_token_subprotocol(&header), Some("tok"));
    }

    #[test]
    fn missing_or_empty_token_decodes_to_none() {
        assert_eq!(decode_token_subprotocol("chat, superchat"), None);
        assert_eq!(decode_token_subprotocol("weldspeak.token."), None);
    }
}
